src/lib.rs

```diff
@@ -1,287 +1,717 @@
+            cubeb_stream_stop(self.native)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn reset_default_device(&self) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_reset_default_device(self.native)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn position(&self) -> Result<u64> {
+        let mut val = 0;
+        let res = unsafe {
+            cubeb_stream_get_position(self.native, &mut val)
+        };
+        match res {
+            CUBEB_OK => Ok(val),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn latency(&self) -> Result<u32> {
+        let mut val = 0;
+        let res = unsafe {
+            cubeb_stream_get_latency(self.native, &mut val)
+        };
+        match res {
+            CUBEB_OK => Ok(val),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn set_volume(&self, volume: f32) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_set_volume(self.native, volume)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn set_panning(&self, panning: f32) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_set_panning(self.native, panning)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+}
+
+impl<T: Sample> Drop for Stream<T> {
+    fn drop(&mut self) {
+        unsafe { cubeb_stream_destroy(self.native); }
+    }
+}
+
+
+pub fn print_state_change(state: State) {
+    match state {
+        State::Started => println!("stream started"),
+        State::Stopped => println!("stream stopped"),
+        State::Drained => println!("stream drained"),
+        State::Error => println!("stream error"),
+    }
+
+}
+
+
+extern fn data_callback_i16(_stm: *const cubeb_stream,
+                            user: *mut c_void,
+                            in_buf: *const c_void,
+                            out_buf: *mut c_void,
+                            nframes: c_long) -> c_long {
     unsafe {
-      if cubeb_get_min_latency(self.get(), params, &mut min_latency) !=
-        CUBEB_OK {
-        println!("failed.");
-      }
-    }
-    min_latency
-  }
-  pub fn preferred_sample_rate(&self) -> u32 {
-    let mut sr : u32 = 0;
+        let data: &mut StreamData<i16> = transmute(user);
+        let ibuf: &[i16] = slice::from_raw_parts(
+            in_buf as *const i16, nframes as usize * data.in_channels as usize
+        );
+        let obuf: &mut [i16] = slice::from_raw_parts_mut(
+            out_buf as *mut i16, nframes as usize * data.out_channels as usize
+        );
+
+        (data.data_cb)(ibuf, obuf) as c_long
+    }
+}
+
+extern fn data_callback_f32(_stm: *const cubeb_stream,
```

Comment:
I was surprised by this code duplication. I'm not knowledgeable enough about rust generics to know if it could be better.

Reply:
I wrote both functions due to the assumption that `extern` implies `#[no_mangle]` and therefore generics would not work. But I think the assumption is wrong, so it can probably work with generics.